internal/kms/kms_ext_test.go

```diff
@@ -0,0 +1,127 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// NOTE: This is a sequential test that relies on the actions that have come
+// before. Please see the comments for details.
+func TestKms(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	conn, _ := db.TestSetup(t, "postgres")
+	rw := db.New(conn)
+	wrapper := db.TestWrapper(t)
+	badWrapper := db.TestWrapper(t)
+	repo, err := kms.NewRepository(rw, rw)
+	require.NoError(t, err)
+	kmsCache := kms.TestKms(t, conn, wrapper)
+	org, proj := iam.TestScopes(t, iam.TestRepo(t, conn, wrapper))
+
+	// Verify that the cache is empty, so we can show that by the end of the
+	// test sequence we did actually look up keys and store them in the cache
+	t.Run("verify cache empty", func(t *testing.T) {
+		var count int
+		kmsCache.GetScopePurposeCache().Range(func(key interface{}, value interface{}) bool {
+			count++
+			return true
+		})
+		assert.Equal(t, 0, count)
+	})
+	// Verify that the root keys are all in the database and can be decrypted
+	// with the correct wrapper from the KMS object
+	t.Run("verify root keys", func(t *testing.T) {
+		assert, require := assert.New(t), require.New(t)
+		rootKeys, err := repo.ListRootKeys(ctx)
+		require.NoError(err)
+		wrappers := kmsCache.GetExternalWrappers()
+		for _, key := range rootKeys {
+			kvs, err := repo.ListRootKeyVersions(ctx, wrappers.Root(), key.GetPrivateId())
+			require.NoError(err)
+			assert.Len(kvs, 1)
+			assert.Len(kvs[0].GetKey(), 32)
+		}
+	})
+	// Verify that the wrong wrapper causes decryption to fail
+	t.Run("bad external keys", func(t *testing.T) {
+		assert, require := assert.New(t), require.New(t)
+		rootKeys, err := repo.ListRootKeys(ctx)
+		require.NoError(err)
+		for _, key := range rootKeys {
+			_, err := repo.ListRootKeyVersions(ctx, badWrapper, key.GetPrivateId())
+			require.Error(err)
+			assert.True(strings.Contains(err.Error(), "message authentication failed"), err.Error())
+		}
+	})
+	// This next sequence is run twice to ensure that calling for the keys twice
+	// returns the same value each time and doesn't simply populate more keys
+	// into the KMS object
+	keyBytes := map[string]bool{}
+	keyIds := map[string]bool{}
+	scopePurposeMap := map[string]interface{}{}
+	for i := 1; i < 3; i++ {
+		// This iterates through wrappers for all three scopes and four purposes,
+		// ensuring that the key bytes and IDs are different for each of them,
+		// simulating calling the KMS object from different scopes for different
+		// purposes and ensuring the keys are different when that happens.
+		t.Run(fmt.Sprintf("verify wrappers different x %d", i), func(t *testing.T) {
+			assert, require := assert.New(t), require.New(t)
+			for _, scopeId := range []string{scope.Global.String(), org.GetPublicId(), proj.GetPublicId()} {
+				for _, purpose := range []kms.KeyPurpose{kms.KeyPurposeUnknown, kms.KeyPurposeOplog, kms.KeyPurposeDatabase, kms.KeyPurposeSessions, kms.KeyPurposeTokens} {
```

Comment:
I'm curious, if we should declare a list/dict of known KeyPurposes?  Just last night I discovered that I didn't add the OIDC dek to `Kms.loadDek(...)` properly.  (I've fixed it in my jimlambrt-oidc-wip branch).  Having a defined list might prevent this sort of issues in the future and would be useful in tests like this as well.

Reply:
I was in process of doing this and I think it's good for tests but wouldn't actually have helped with the `loadDek` issue. The issue there really is that the switch case needs a `default`.